controllers/nodehealthcheck_controller.go

```diff
@@ -160,174 +195,101 @@ func (r *NodeHealthCheckReconciler) Reconcile(ctx context.Context, req ctrl.Requ
-			// some actors want to pause remediation.
-			msg := "Skipping remediation because there are pause requests"
-			log.Info(msg)
-			r.Recorder.Event(nhc, eventTypeNormal, eventReasonRemediationSkipped, msg)
-			return false
+	for _, node := range unhealthyNodes {
+		remediationCR := resourceManager.GenerateRemediationCR(&node, nhc, template)
+		nextReconcile, err := r.remediate(&node, nhc, remediationCR, resourceManager)
+		if err != nil {
+			// don't try to remediate other nodes
+			log.Error(err, "failed to start remediation")
+			return result, err
 		}
-		// TODO consider doing this check on top of reconcile and set Disabled condition?
-		if r.isClusterUpgrading() {
-			updateResultNextReconcile(result, clusterUpgradeRequeueAfter)
-			r.Recorder.Event(nhc, eventTypeNormal, eventReasonRemediationSkipped, "Skipped remediation because the cluster is upgrading")
-			return false
+		if nextReconcile != nil {
+			updateResultNextReconcile(&result, *nextReconcile)
 		}
-		return true
 	}
-	msg := fmt.Sprintf("Skipped remediation because the number of healthy nodes selected by the selector is %d and should equal or exceed %d", healthyNodes, minHealthy)
-	log.Info(msg, "healthyNodes", healthyNodes, "minHealthy", minHealthy)
-	r.Recorder.Event(nhc, eventTypeWarning, eventReasonRemediationSkipped, msg)
-	return false
+
+	return result, nil
 }
 
 func (r *NodeHealthCheckReconciler) isClusterUpgrading() bool {
 	clusterUpgrading, err := r.ClusterUpgradeStatusChecker.Check()
 	if err != nil {
-		// log the error but don't return - if we can't reliably tell if
-		// the cluster is upgrading then just continue with remediation.
+		// if we can't reliably tell if the cluster is upgrading then just continue with remediation.
 		// TODO finer error handling may help to decide otherwise here.
 		r.Log.Error(err, "failed to check if the cluster is upgrading. Proceed with remediation as if it is not upgrading")
 	}
-	if clusterUpgrading {
-		r.Log.Info("Skipping remediation because the cluster is currently upgrading.")
-		return true
-	}
-	return false
-}
-
-func (r *NodeHealthCheckReconciler) fetchNodes(ctx context.Context, labelSelector metav1.LabelSelector) ([]v1.Node, error) {
-	var nodes v1.NodeList
-	selector, err := metav1.LabelSelectorAsSelector(&labelSelector)
-	if err != nil {
-		err = errors.Wrapf(err, "failed converting a selector from NHC selector")
-		return []v1.Node{}, err
-	}
-	err = r.List(
-		ctx,
-		&nodes,
-		&client.ListOptions{LabelSelector: selector},
-	)
-	return nodes.Items, err
+	return clusterUpgrading
 }
 
-func (r *NodeHealthCheckReconciler) checkNodesHealth(nodes []v1.Node, nhc *remediationv1alpha1.NodeHealthCheck, template *unstructured.Unstructured) ([]v1.Node, error) {
-	var unhealthy []v1.Node
-	for i := range nodes {
-		node := &nodes[i]
+func (r *NodeHealthCheckReconciler) checkNodesHealth(nodes []v1.Node, nhc *remediationv1alpha1.NodeHealthCheck) (healthy []v1.Node, unhealthy []v1.Node) {
```

Comment:
By giving names to output variables we also set them? IDK that :)
I didn't find any declaration of `healthy` and `unhealthy` variables beforehand :thinking

Reply:
correct

Reply:
Nice